elephant/surrogates.py

```diff
@@ -0,0 +1,490 @@
+import neo
+import elephant.statistics as es
+
+
+def spike_dithering(spiketrain, dither, n=1, decimals=None, edges=True):
+    """
+    Generates surrogates of a spike train by spike dithering.
+
+    The surrogates are obtained by uniformly dithering times around the
+    original position. The dithering is performed independently for each
+    surrogate.
+
+    The surrogates retain the :attr:`t_start` and :attr:`t_stop` of the
+    original `SpikeTrain` object. Spikes moved beyond this range are lost or
+    moved to the range's ends, depending on the parameter edge.
+
+
+    Parameters
+    ----------
+    spiketrain :  neo.SpikeTrain
+        The spike train from which to generate the surrogates
+    dither : quantities.Quantity
+        Amount of dithering. A spike at time t is placed randomly within
+        ]t-dither, t+dither[.
+    n : int (optional)
+        Number of surrogates to be generated.
+        Default: 1
+    decimals : int or None (optional)
+        Number of decimal points for every spike time in the surrogates
+        If None, machine precision is used.
+        Default: None
+    edges : bool (optional)
+        For surrogate spikes falling outside the range
+        `[spiketrain.t_start, spiketrain.t_stop)`, whether to drop them out
+        (for edges = True) or set that to the range's closest end
+         (for edges = False).
+        Default: True
+
+    Returns
+    -------
+    list of neo.SpikeTrain
+      A list of `neo.SpikeTrain`, each obtained from :attr:`spiketrain` by
+      randomly dithering its spikes. The range of the surrogate spike trains
+      is the same as :attr:`spiketrain`.
+
+    Example
+    -------
+    >>> import quantities as pq
+    >>> import neo
+    >>>
+    >>> st = neo.SpikeTrain([100, 250, 600, 800]*pq.ms, t_stop=1*pq.s)
+    >>> print spike_dithering(st, dither = 20*pq.ms)
+    [<SpikeTrain(array([  96.53801903,  248.57047376,  601.48865767,
+     815.67209811]) * ms, [0.0 ms, 1000.0 ms])>]
+    >>> print spike_dithering(st, dither = 20*pq.ms, n=2)
+    [<SpikeTrain(array([ 104.24942044,  246.0317873 ,  584.55938657,
+        818.84446913]) * ms, [0.0 ms, 1000.0 ms])>,
+     <SpikeTrain(array([ 111.36693058,  235.15750163,  618.87388515,
+        786.1807108 ]) * ms, [0.0 ms, 1000.0 ms])>]
+    >>> print spike_dithering(st, dither = 20*pq.ms, decimals=0)
+    [<SpikeTrain(array([  81.,  242.,  595.,  799.]) * ms,
+        [0.0 ms, 1000.0 ms])>]
+    """
+
+    # Transform spiketrain into a Quantity object (needed for matrix algebra)
+    data = spiketrain.view(pq.Quantity)
+
+    # Main: generate the surrogates
+    surr = data.reshape((1, len(data))) + 2 * dither * np.random.random_sample(
+        (n, len(data))) - dither
+
+    # Round the surrogate data to decimal position, if requested
+    if decimals is not None:
+        surr = surr.round(decimals)
+
+    if edges is False:
+        # Move all spikes outside [spiketrain.t_start, spiketrain.t_stop] to
```

Comment:
It might be done if one then needs to bin the surrogate spiketrain into a 0-1 sequence, or wants to preserve the total spike count. I am personally fine with dropping the option though - never used it.
different t_start and t_stop are not fine instead, because of the meaning of surrogate data - stochastic spike times inside a well-defined and fixed time window [t_start,t_stop].